docs: document the customisation flow (#88)
asyncapi/generator-model-sdk: docs/customisation.md

# Customisation

> The AsyncAPI Model SDK uses **presets** objects to extend the rendered model.

Reviewer: ```suggestion 
The AsyncAPI Model SDK uses **preset** objects to extend the rendered model.
```

## Preset

**Preset** is a pure JavaScript object with format `key: value`, where `key` is a name of model type and `value` is the object which contains methods, which extend a given rendered part for a given model type, like below example:

```js
{
  // `class` model type 
  class: {
    self(...options) { /* logic */ },
    // `setter` customization method 
    setter(...options) { /* logic */ },
  },
  interface: {
    // `property` customization method 
    property(...options) { /* logic */ },
    additionalContent(...options) { /* logic */ },
  },
}
```

Each language has different model types, which results in different implementable methods in a single preset. For more information, please check the [preset's shape](#presets-shape) section.

## Custom preset

> Below is a custom preset written for TypeScript language, which adds a description to each interface's property and to the self as a JavaScript comment.

Reviewer 2: ```suggestion
Below is a custom preset written for TypeScript language, which adds a description to each interface's property and to the model itself as a JavaScript comment.
```

The user has the possibility to pass options to the custom preset - if preset has defined options - used in a `presets` array. To do that, please pass as an array's element an object with the `preset` field as a reference to the custom preset and define the `options` field with options passed to the preset. Check example:

```js
const generator = new TypeScriptGenerator({ presets: [
  SOME_PRESET, 
  {
    preset: PRESET_WITH_OPTIONS,
    options: {...},
  },
] });
```

## Overriding the default preset

Each implemented generator for appropriate language must have defined the default preset. However, we can override it by passing it as the `defaultPreset` parameter in the generator options. Check the example for TypeScript generator:

```js
const DEFAULT_PRESET = {
  // implementation
}

const generator = new TypeScriptGenerator({ defaultPreset: DEFAULT_PRESET });
```

> **NOTE**: Each default preset must have implemented each method for a given model type. Keep this in mind when overriding the default preset.

## Preset's shape

> By each preset user can implement two basic methods:

Author: Yes, this is true, because each model type, like class, enum, interface has by default these two method and then appropriate model - like class - has another methods like setter, getter etc.

I'll change this sentence to `For each model type, you can implement two basic methods:`

> - `self` - the method for extending the whole model shape.

Author: I wanna leave the `extending` word, because as in example showed in document you can also extending model shape, not only replacing.

> - `self` - the method for extending the whole model shape.

Author: So I removed the `whole` word, it causes misunderstanding.

> - `self` - the method for extending the whole model shape.

Reviewer 2: Yeah, much better. Thanks!

- `additionalContent` - the method which adds additional content to the model.

Each customization method receives the following arguments:

- `model` - an instance of the [`CommonModel`](../src/models/CommonModel.ts) class, which described rendered data model.
- `inputModel` - an instance of the [`CommonInputModel`](../src/models/CommonInputModel.ts) class.
- `renderer` - an instance of the class with common helper functions to render appropriate model type.

> - `content` - rendered content from previous preset.

Reviewer: ```suggestion 
- `content` - rendered content based on the language and preset.
```
Maybe?

Author: You cannot change language in rendering process and `content` is exactly rendered content from previous preset.

| `setter` | A method to extend setter for a given property. | `propertyName` as a name of a given property, `property` object as a [`CommonModel`](../src/models/CommonModel.ts) instance and `parentModel` as a [`CommonModel`](../src/models/CommonModel.ts) instance. |
| `getter` | A method to extend getter for a given property. | `propertyName` as a name of a given property, `property` object as a [`CommonModel`](../src/models/CommonModel.ts) instance and `parentModel` as a [`CommonModel`](../src/models/CommonModel.ts) instance. |

#### **Interface**

| Method | Description | Additional arguments |
|---|---|---|
| `property` | A method to extend rendered given property. | `propertyName` as a name of a given property, `property` object as a [`CommonModel`](../src/models/CommonModel.ts) instance and `parentModel` as a [`CommonModel`](../src/models/CommonModel.ts) instance. |

#### **Enum**

| Method | Description | Additional arguments |
|---|---|---|
| `item` | A method to extend enum's item. | an `item` containing the value of enum's item and `parentModel` as a [`CommonModel`](../src/models/CommonModel.ts) instance. |

#### **Type**

> There are no additional methods.

Reviewer: I am just gonna put the same comment @smoya left for my PR regarding quick outdated docs. Can we add jsdoc to make these explanations?

Author: I also thought about that, but the problem is that types for presets are a TS interfaces. Also another options was pointing to source code of preset types, but there we have a problem, because types use generics, so it hard to read if you are not familiar with TS. In the future I have a plan for each generator to have a separate document with options/presets etc and then it will be easier to maintain.

Author: I tried to add jsdoc to interfaces and it doesn't work.

Reviewer 3: Totally unfamiliar with TS jsdoc but, are we sure it is not a supported feature? I saw some issues where seems it's a thing - E.g. https://github.com/microsoft/TypeScript/issues/18358

Author: Yes, TS has supports for jsdoc, but interfaces is missed after transpilation to JS and even if I changed script related to render the `API.md` to `./node_modules/.bin/jsdoc2md lib/index.js -f lib/**/*.ts > API.md` then I have empty file.

Author: We transpile TS to JS and from JS we render docs.

Reviewer 3: I guess that's why things like https://github.com/SoftwareBrothers/better-docs#how-it-works exist. Also, you can generate docs from TS from it afaik. Not sure what's the real effort of this, but I guess it's a topic non purely related with this PR.

Author: We use https://github.com/jsdoc2md/jsdoc-to-markdown so at the moment we don't want to switch :)

Reviewer 2: @smoya maybe you can create an issue so we don't forget about better-docs package.

Reviewer 3: > We use https://github.com/jsdoc2md/jsdoc-to-markdown so at the moment we don't want to switch :)
Fair enough 👍 

> @smoya maybe you can create an issue so we don't forget about better-docs package.

Yup, makes totally sense. Thanks for pointing that out @fmvilas

Author: @smoya Maybe only in https://github.com/asyncapi/generator-react-sdk. In other place we use pure JS.

Reviewer 2: Let's just leave it there and if we prefer this one over the jsdoc-to-markdown maybe we can just replace it everywhere.
